ComponentKit/Utilities/CKComponentAction.mm

```diff
@@ -209,42 +210,34 @@ CKComponentViewAttributeValue CKComponentActionAttribute(const CKTypedComponentA
     };
   }
 
-  // We need a target for the control event. (We can't use the responder chain because we need to jump in and change the
-  // sender from the UIControl to the CKComponent.)
-  // Control event targets are __unsafe_unretained. We can't rely on the block to keep the target alive, since the block
-  // is merely an "applicator"; if the attributes compare the same (say, two equivalent attributes used across two
-  // versions of the same component) then the block may be deallocated on the first one without removing the attribute.
-  // Thus we create a map from component action to forwarders and never release the forwarders.
-  // If this turns out to have memory overhead, we could capture a "token" in the blocks and have those tokens as ref-
-  // counts on the forwarder, and when the number of outstanding tokens goes to zero, release the forwarder.
-  // However I expect the number of actions to be O(200) at most and so the memory overhead is not a concern.
-  CKComponentActionControlForwarder *forwarder;
-  {
-    CK::StaticMutexLocker l(lock);
-    auto it = map->find(action);
-    if (it == map->end()) {
-      forwarder = [[CKComponentActionControlForwarder alloc] initWithAction:action];
-      map->insert({action, forwarder});
-    } else {
-      forwarder = it->second;
-    }
-  }
-
-  std::string identifier = std::string("CKComponentActionAttribute-")
-  + action.identifier()
-  + "-" + std::to_string(controlEvents);
+  static ForwarderMap *map = new ForwarderMap(); // access on main thread only; never destructed to avoid static destruction fiasco
   return {
     {
-      identifier,
+      std::string("CKComponentActionAttribute-") + action.identifier() + "-" + std::to_string(controlEvents),
       ^(UIControl *control, id value){
-        [control addTarget:forwarder
-                    action:@selector(handleControlEventFromSender:withEvent:)
-          forControlEvents:controlEvents];
+        CKComponentActionList *list = objc_getAssociatedObject(control, ck_actionListKey);
+        if (list == nil) {
+          list = [CKComponentActionList new];
+          objc_setAssociatedObject(control, ck_actionListKey, list, OBJC_ASSOCIATION_RETAIN_NONATOMIC);
+
+          // Since this is the first time we've seen this control, add a Forwarder as a target.
+          const auto it = map->find(controlEvents);
+          CKComponentActionControlForwarder *const forwarder =
+          (it == map->end())
+          ? map->insert({controlEvents, [[CKComponentActionControlForwarder alloc] initWithControlEvents:controlEvents]}).first->second
+          : it->second;
+          [control addTarget:forwarder
+                      action:@selector(handleControlEventFromSender:withEvent:)
+            forControlEvents:controlEvents];
+        }
+        list->_actions[controlEvents].push_back(action);
```

Comment:
Actually I did mean "if it never runs." There is no guarantee that an unapplicator will run.